src/katsdpcontroller/scheduler.py

```diff
@@ -3778,6 +3790,92 @@ def set_driver(self, driver):
+                    flags=socket.AI_ADDRCONFIG | socket.AI_V4MAPPED,
+                )
+                # getaddrinfo always returns at least 1 (it is an error if there are no
+                # matches), so we do not need to check for the empty case.
+                return addrs[0][4][0]  # First entry, sockaddr, address
+            except socket.gaierror as error:
+                if retries > 0:
+                    logger.warning(
+                        "Failure to resolve address for %s (%s). Waiting 1s to retry.",
+                        hostname,
+                        error,
+                    )
+                    await asyncio.sleep(1)
+                else:
+                    raise
+
+    async def _resolve_offer(self, offer):
+        """Resolve the hostname of an offer to an IP address and record the offer."""
+        offer.address = await self._resolve_hostname(offer.hostname)
+        role = offer.allocation_info.role
+        if role in self._roles_needed:
+            logger.debug(
+                "Adding offer %s on %s (%s) with role %s to pool",
+                offer.id.value,
+                offer.agent_id.value,
+                offer.hostname,
+                role,
+            )
+            role_offers = self._offers.setdefault(role, {})
+            agent_offers = role_offers.setdefault(offer.agent_id.value, {})
+            agent_offers[offer.id.value] = offer
+            self._wakeup_launcher.set()
+        else:  # pragma: nocover
+            # This should be unreachable, because the role was in _roles_needed
+            # when we kicked off _resolve_offer, and when it is removed (in
+            # _update_roles) we cancel this task.
+            logger.warning(
+                "Offer %s on %s (%s) with role %s is no longer needed (possible bug)",
+                offer.id.value,
+                offer.agent_id.value,
+                offer.hostname,
+                role,
+            )
+            self._driver.declineOffer([offer.id])
+
+    def _resolve_offer_done(self, resolving_offer: ResolvingOffer, task: asyncio.Task) -> None:
+        """Clean up after :meth:`_resolve_offer`.
+
+        This is added as a done callback rather than a try block inside
+        :meth:`_resolve_offer` to guarantee that it will be used even if
+        the task is cancelled before it gets a chance to run at all.
+        """
+        decline = not resolving_offer.rescinded
+        offer = resolving_offer.offer
+        try:
+            task.result()  # Trigger exception, if any
+        except asyncio.CancelledError:
+            pass
+        except Exception:
+            logger.warning(
+                "Failed to resolve %s for offer %s",
+                offer.hostname,
+                offer.id.value,
+                exc_info=True,
+            )
+        else:
+            # If we get there, the resolution succeeded, and the offer is no
+            # longer our responsibility.
```

Comment:
Or we hit the else path in _resolve_offer, in which case _resolve_offer already declined and we don't need to do it again.